From b57b24e0affbad5ac62821d3ee5145c061489c9d Mon Sep 17 00:00:00 2001
From: Marcin Wojtas <mw@semihalf.com>
Date: Tue, 5 Feb 2019 09:48:48 +0100
Subject: [PATCH 0994/1051] net: mvneta: add branch prediction for BM usage in
 RX hot-path

Usage of HW buffer manager (BM) is configurable per-port and
can change in runtime, e.g. upon change MTU fail. Because of that
a static_branch array should be used for each port separately,
in order to properly set branch prediction in mvneta_poll routine.

Unfortunately due to generic issue, index of an array initialized
with DEFINE_STATIC_KEY_ARRAY_FALSE/TRUE must be a local variable
- using global value (BSS) or structure field (heap) is illegal
and cause compilation failure.

However it is still worth in terms of performance to fix branch
prediction for 64-bit variant, which does not support hardware
buffer manager.

Change-Id: I03826f4815617e385ed284f52ec42fe2102e44c7
Signed-off-by: Marcin Wojtas <mw@semihalf.com>
Reviewed-on: https://sj1git1.cavium.com/3750
Tested-by: sa_ip-sw-jenkins
Reviewed-by: Stefan Chulski <Stefan.Chulski@cavium.com>
[Kevin: The original patch got from Marvell sdk10.0_19.06]
Signed-off-by: Kevin Hao <kexin.hao@windriver.com>
---
 drivers/net/ethernet/marvell/mvneta.c | 8 ++++++++
 1 file changed, 8 insertions(+)

diff --git a/drivers/net/ethernet/marvell/mvneta.c b/drivers/net/ethernet/marvell/mvneta.c
index 54aebfb608e8..71488c3a03c8 100644
--- a/drivers/net/ethernet/marvell/mvneta.c
+++ b/drivers/net/ethernet/marvell/mvneta.c
@@ -2819,7 +2819,15 @@ static int mvneta_poll(struct napi_struct *napi, int budget)
 
 	if (rx_queue) {
 		rx_queue = rx_queue - 1;
+#if defined(CONFIG_ARM64)
+		/* Hardcode branch prediction for 64-bit variant,
+		 * which is a workaround for generic issues with using
+		 * static_branch array.
+		 */
+		if (unlikely(pp->bm_priv))
+#else
 		if (pp->bm_priv)
+#endif
 			rx_done = mvneta_rx_hwbm(napi, pp, budget,
 						 &pp->rxqs[rx_queue]);
 		else
-- 
2.17.1

